File: final_project/symnmf.py
import sys
import numpy as np
import pandas as pd
import symnmf as sm


def parse_int(num):
    try:
        return int(float(num)) if int(float(num)) == float(num) else 0
    except:
        return 0


def parse_float(num):
    try:
        return float(num)
    except:
        return -1


np.random.seed(1234)

error_msg = "An Error Has Occurred"
max_iter = 300
eps = 0.0001

if len(sys.argv) < 4:
    print(error_msg)
    sys.exit(1)

try:
    k = parse_int(sys.argv[1])
    goal = sys.argv[2]
    file_name = sys.argv[3]

    points = pd.read_csv(file_name, header=None)
    n = len(points)

    if not (1 < k < n):
        sys.exit(1)

    # A = np.fromfunction(
    #     lambda i, j: (
    #         np.exp(-(np.pow(np.linalg.norm(points.iloc[i] - points.iloc[j]), 2) / 2))
    #         if i != j
    #         else 0
    #     ),
    #     (n, n),
    # )
    #     D_inv_sqrt = np.diag([np.sum(A[i]) for i in range(n)])
    # D = sm.ddg(X)

    X = points.values.tolist()
    H = np.random.uniform(0, 2 * np.sqrt(m / k), (n, k))
    W = sm.norm(X)

    # TODO check Nova's python version
    match goal:
        case "symnmf":
            res = sm.symnmf(H, W)
        case "sym":
            res = sm.sym(X)
        case "ddg":
            res = sm.ddg(X)
        case "norm":
            res = sm.norm(X)
        case _:
            res = []

    for line in res:
        print(",".join(["%.4f" % coord for coord in line]))

except:
    print(error_msg)
    sys.exit(1)

File: final_project/symnmf.c
#include "symnmf.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void free_rec(double **mat, int n) {
    int i;
    if (mat == NULL) return;
    for (i = 0; i < n; ++i) free(mat[i]);
    free(mat);
}

double euc_dist_sq(double *point1, double *point2, int dim) {
    double sum = 0;
    int i;
    for (i = 0; i < dim; ++i) sum += pow(point1[i] - point2[i], 2);
    return sum;
}

/** multiply n*k matrix mat1 with k*m matrix mat2 */
double **mat_mult(double **mat1, double **mat2, int n, int k, int m) {
    double **res;
    int i, j, t;

    res = calloc(n, sizeof(double *));
    if (res == NULL) return NULL;
    for (i = 0; i < n; ++i) {
        res[i] = calloc(m, sizeof(double));
        if (res[i] == NULL) {
            free_rec(res, n);
            return NULL;
        }
    }

    for (i = 0; i < n; ++i) {
        for (j = 0; j < m; ++j) {
            double sum = 0;
            for (t = 0; t < k; ++t) sum += mat1[i][t] * mat2[t][j];
            res[i][j] = sum;
        }
    }

    return res;
}

double **sym(double **x, int n, int dim) {
    double **a;
    int i, j;

    a = calloc(n, sizeof(double *));
    if (a == NULL) return NULL;
    for (i = 0; i < n; ++i) {
        a[i] = calloc(n, sizeof(double));
        if (a[i] == NULL) {
            free_rec(a, n);
            return NULL;
        }
    }

    for (i = 0; i < n; ++i) {
        for (j = 0; j < n; ++j) {
            if (i != j)
                a[i][j] = exp(-(euc_dist_sq(x[i], x[j], dim) / 2));
        }
    }

    return a;
}

double **ddg(double **x, int n, int dim) {
    double **a, **d;
    int i, j;

    a = sym(x, n, dim);
    if (a == NULL) return NULL;

    d = calloc(n, sizeof(double *));
    if (d == NULL) {
        free_rec(a, n);
        return NULL;
    }
    for (i = 0; i < n; ++i) {
        d[i] = calloc(n, sizeof(double));
        if (d[i] == NULL) {
            free_rec(a, n);
            free_rec(d, n);
            return NULL;
        }
    }

    for (i = 0; i < n; ++i) {
        double sum = 0;
        for (j = 0; j < dim; ++j) sum += a[i][j];
        d[i][i] = sum;
    }

    free_rec(a, n);

    return d;
}

double **norm(double **x, int n, int dim) {
    double **a, **d, **tmp_w, **w;
    int i;

    a = sym(x, n, dim);
    if (a == NULL) return NULL;
    d = ddg(x, n, dim);
    if (d == NULL) {
        free_rec(a, n);
        return NULL;
    }

    w = calloc(n, sizeof(double *));
    if (w == NULL) {
        free_rec(a, n);
        free_rec(d, n);
        return NULL;
    }
    for (i = 0; i < n; ++i) {
        w[i] = calloc(n, sizeof(double));
        if (w[i] == NULL) {
            free_rec(a, n);
            free_rec(d, n);
            free_rec(w, n);
            return NULL;
        }
    }

    for (i = 0; i < n; ++i) d[i][i] = sqrt(1 / d[i][i]);
    tmp_w = mat_mult(d, a, n, n, n);
    w = mat_mult(tmp_w, d, n, n, n);

    free_rec(a, n);
    free_rec(d, n);
    free_rec(tmp_w, n);

    return w;
}
